[FX][RFC] Input prototype for collection types in sym trace

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43741 [FX][RFC] Input prototype for collection types in sym trace**
* #43740 [FX] Better error when unpacking Proxy
* #43674 [FX] Pickle serialization of GraphModule via forward source
* #43666 [FX] Support tensor-valued constants

A stab at creating an API to alleviate https://github.com/pytorch/pytorch/issues/43679. This defines `input_prototype` as an argument to `symbolic_trace`. This can be used to specify--for each function argument for the fn being traced--what is the structure of the argument if it is a collection type. This exposes `SymVal` as a placeholder for "leaf" values in the collection type that should be left as `Proxy`s.

Differential Revision: [D23382444](https://our.internmc.facebook.com/intern/diff/D23382444)

Commit:
[FX][RFC] Input prototype for collection types in sym trace

[ghstack-poisoned]

Commit:
Update on "[FX][RFC] Input prototype for collection types in sym trace"


A stab at creating an API to alleviate https://github.com/pytorch/pytorch/issues/43679. This defines `input_prototype` as an argument to `symbolic_trace`. This can be used to specify--for each function argument for the fn being traced--what is the structure of the argument if it is a collection type. This exposes `SymVal` as a placeholder for "leaf" values in the collection type that should be left as `Proxy`s.

Differential Revision: [D23382444](https://our.internmc.facebook.com/intern/diff/D23382444)

[ghstack-poisoned]

dzhulgakov commented on torch/fx/symbolic_trace.py:
```diff
@@ -151,6 +151,21 @@ def search_for_tensor(m : torch.nn.Module, qualname_atoms : List[str]) -> Option
+    pass
+
+def unpack_collection_type(proxy : Proxy, prototype : Any, delegate : DelegateBase):
+    if isinstance(prototype, list):
```
also add Tuple?

dzhulgakov commented on torch/fx/symbolic_trace.py:
```diff
@@ -151,6 +151,21 @@ def search_for_tensor(m : torch.nn.Module, qualname_atoms : List[str]) -> Option
+def unpack_collection_type(proxy : Proxy, prototype : Any, delegate : DelegateBase):
+    if isinstance(prototype, list):
+        return [unpack_collection_type(proxy[i], prototype[i], delegate) for i in range(len(prototype))]
+    if isinstance(prototype, dict):
```
I wonder whether we should generate some form of assert that the keys are always the same (and for lists that the lengths is constant). Not sure how represent those asserts in IR though

dzhulgakov commented on torch/fx/symbolic_trace.py:
```diff
@@ -172,6 +189,12 @@ def symbolic_trace(root : torch.nn.Module, delegate_class=DefaultDelegate) -> Gr
     args.extend(_proxy_placeholder(next(names_iter), delegate) for name in range(1, total_args))
 
+    if input_prototypes:
+        assert isinstance(input_prototypes, list)
```
list or tuple?

Commit:
Update on "[FX][RFC] Input prototype for collection types in sym trace"


A stab at creating an API to alleviate https://github.com/pytorch/pytorch/issues/43679. This defines `input_prototype` as an argument to `symbolic_trace`. This can be used to specify--for each function argument for the fn being traced--what is the structure of the argument if it is a collection type. This exposes `SymVal` as a placeholder for "leaf" values in the collection type that should be left as `Proxy`s.

Differential Revision: [D23382444](https://our.internmc.facebook.com/intern/diff/D23382444)

[ghstack-poisoned]

Commit:
Update on "[FX][RFC] Input prototype for collection types in sym trace"


A stab at creating an API to alleviate https://github.com/pytorch/pytorch/issues/43679. This defines `input_prototype` as an argument to `symbolic_trace`. This can be used to specify--for each function argument for the fn being traced--what is the structure of the argument if it is a collection type. This exposes `SymVal` as a placeholder for "leaf" values in the collection type that should be left as `Proxy`s.

Differential Revision: [D23382444](https://our.internmc.facebook.com/intern/diff/D23382444)

[ghstack-poisoned]

Commit:
Update on "[FX][RFC] Input prototype for collection types in sym trace"


A stab at creating an API to alleviate https://github.com/pytorch/pytorch/issues/43679. This defines `input_prototype` as an argument to `symbolic_trace`. This can be used to specify--for each function argument for the fn being traced--what is the structure of the argument if it is a collection type. This exposes `SymVal` as a placeholder for "leaf" values in the collection type that should be left as `Proxy`s.

Differential Revision: [D23382444](https://our.internmc.facebook.com/intern/diff/D23382444)

[ghstack-poisoned]

Commit:
Update on "[FX][RFC] Input prototype for collection types in sym trace"


A stab at creating an API to alleviate https://github.com/pytorch/pytorch/issues/43679. This defines `input_prototype` as an argument to `symbolic_trace`. This can be used to specify--for each function argument for the fn being traced--what is the structure of the argument if it is a collection type. This exposes `SymVal` as a placeholder for "leaf" values in the collection type that should be left as `Proxy`s.

Differential Revision: [D23382444](https://our.internmc.facebook.com/intern/diff/D23382444)

[ghstack-poisoned]

Commit:
Update on "[FX][RFC] Input prototype for collection types in sym trace"


A stab at creating an API to alleviate https://github.com/pytorch/pytorch/issues/43679. This defines `input_prototype` as an argument to `symbolic_trace`. This can be used to specify--for each function argument for the fn being traced--what is the structure of the argument if it is a collection type. This exposes `SymVal` as a placeholder for "leaf" values in the collection type that should be left as `Proxy`s.

Differential Revision: [D23382444](https://our.internmc.facebook.com/intern/diff/D23382444)

[ghstack-poisoned]

Commit:
Update on "[FX][RFC] Input prototype for collection types in sym trace"


A stab at creating an API to alleviate https://github.com/pytorch/pytorch/issues/43679. This defines `input_prototype` as an argument to `symbolic_trace`. This can be used to specify--for each function argument for the fn being traced--what is the structure of the argument if it is a collection type. This exposes `SymVal` as a placeholder for "leaf" values in the collection type that should be left as `Proxy`s.

Differential Revision: [D23382444](https://our.internmc.facebook.com/intern/diff/D23382444)

[ghstack-poisoned]

Commit:
Update on "[FX][RFC] Input prototype for collection types in sym trace"


A stab at creating an API to alleviate https://github.com/pytorch/pytorch/issues/43679. This defines `input_prototype` as an argument to `symbolic_trace`. This can be used to specify--for each function argument for the fn being traced--what is the structure of the argument if it is a collection type. This exposes `SymVal` as a placeholder for "leaf" values in the collection type that should be left as `Proxy`s.

Differential Revision: [D23382444](https://our.internmc.facebook.com/intern/diff/D23382444)

[ghstack-poisoned]
